byway/src/main/java/com/google/sps/servlets/EntityServlet.java

```diff
@@ -0,0 +1,115 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceConfig;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.gson.Gson;
+import com.google.sps.data.Trip;
+import com.google.sps.data.UserInfo;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/entity")
+public class EntityServlet extends HttpServlet {
+  private final Gson gson = new Gson();
+  private final UserService userService = UserServiceFactory.getUserService();
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    com.google.appengine.api.users.User user = userService.getCurrentUser();
+    UserInfo userInfo = addUserEntity(user);
```

Comment:
Remember that userService.getCurrentUser() might return `null`. The way I would recommend handling that would be to have the static method `UserInfo.findOrCreateUser(UserService userService)` on the `UserInfo` class handle calling `getCurrentUser` and return a `null` value for `UserInfo` if the `UserService` returns a null value for `getCurrentUser`. That way it's a single function call to retrieve user info from datastore, and then in the calling code you can just do a null-check against the returned `UserInfo` object.

You can then do a login-check with something like:

```java
UserInfo user = UserInfo.findOrCreateUser(userService);
if (user == null) {
  response.setStatus(HttpServletResponse.SC_UNAUTHORIZED)
}
```

Note that I would use SC_UNAUTHORIZED to indicate that the user is not logged in, and SC_FORBIDDEN if a user tries to access a trip that doesn't belong to them.

Reply:
done